### PR TITLE
Note input improvements

### PR DESCRIPTION
**Pull request by dgarrett**
As a recent Koreader convert, I've found that Koreader's current note input UI makes it difficult to input my own longer notes. I found some tweaks that bring the note input behavior more in line with other ebook UIs.

- Display auto-text as hint-text so it doesn't have to be manually cleared to replace it with a personal note
- Default to a larger text box to allow for nicer multi-line notes
- Default cursor to the end of the note to allow for easier appending

I'm not sure how much interest there is in this kind of change. Please let me know if there's a preferred way to provide this kind of note input behavior. I'm glad to iterate on this.

Auto-text is displayed as hint-text, so the user doesn't need to delete it manually if they want to enter their own note:
Screen Shot 2021-11-21 at 6 52 31 PM

Can display 5 lines at default DPI:
Screen Shot 2021-11-21 at 7 03 41 PM

UI immediately after opening a six-line note:
Screen Shot 2021-11-21 at 6 54 21 PM

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/8467)

### CODE REVIEW

**Comment by hius07** [comment]
> Auto-text is displayed as hint-text, so the user doesn't need to delete it manually if they want to enter their own note:

What to do if I want a part of highlighted text to be included into the note (a frequent use case)?
Deleting the full note is quick - just long-press the backspace.



> Can display 5 lines at default DPI:

Proposed 120 height displays fixed 2 lines only on my Kindle.
Current flexible window is scaled for long highlights.

**Comment by poire-z** [comment]
Dunno why I initially set `cursor_at_end = false`, may be it was the original behaviour and I didn't dare change it. `= true` feels indeed more logical: you want to append, or long-press backspace to erase it. But note that in the snippet below, if overflow, we put the cursor at start - may be I did it so this is consistent, whether the text is short or overflowing ?

`text_height = 120,` as a fixed unscaled for DPI size, shouldn't be used. It should either be a % of the remaining space above the keyboard, or (to implemented in InputDialog) a nb of lines we want to show (which should still be caped to what can fit above the keyboard.

Or just have an other option in InputDialog to use this remaining space always, and not only when the text is bigger.
https://github.com/koreader/koreader/blob/dcd34d409e6e3c3c73ced9dbf5d6996e66c657ec/frontend/ui/widget/inputdialog.lua#L316-L364

**Comment by dgarrett** [comment]
Thanks for the feedback. And thanks for testing it out on your Kindle, @hius07.

I was starting from how I've used the Kobo and Kindle official reader applications, which start you out with a blank note. I can see how starting out with the highlighted text pre-filled can be useful.  I'll revert back to the original behavior of `input`. Though it would be much nicer for the case of wanting to start from a blank note with `cursor_at_end = true`, which I'd really like to leave for easy backspacing. I think the overflow behavior is still reasonable, since it lets you see the beginning of a long note.

Good points about not using a fixed pixel size and I clearly missed important testing scenarios. Let me look into exposing an option on InputDialog to always use available space. I really don't like how the available space for a note is currently limited based on the length of the highlight itself. A short highlight doesn't mean my note will be short.

By the way, what are general thoughts on adding options? For example, if I were to add an option for notes to start out blank?

**Comment by hius07** [comment]
Agree with cursor at the end and bigger input box.
Doubt of a new setting. Another way is to start a note with blank input and have highlighted text available via clipboard. Showing a hint, something like "Highlighted text has been copied to clipboard. Long-press to paste".

**Comment by dgarrett** [comment]
I like the clipboard idea. It seems to be more of a functional change than the cursor at the end and bigger input box, so I'll follow up with another PR for that when I get the chance.

**Comment by dgarrett** [comment]
I pushed an update with the `use_available_height` idea.

Opening a comment to edit that fits on the screen:
Screen Shot 2021-11-25 at 6 09 19 PM

Starting a note on a long highlight. The overflow behavior still exists and the cursor starts at the beginning:
Screen Shot 2021-11-25 at 6 09 00 PM

File: frontend/ui/widget/inputdialog.lua [hunk 1]
```diff
@@ -142,6 +142,7 @@ local InputDialog = InputContainer:new{
                              -- needs add_nav_bar to have a Show keyboard button to get it back
     scroll_by_pan = false, -- allow scrolling by lines with Pan (= Swipe, but wait a bit at end
                            -- of gesture before releasing) (may conflict with movable)
+    use_available_height = false, -- adjust input box to fill available height on screen
```
Comment (poire-z) [inline frontend/ui/widget/inputdialog.lua:145]: May be put it nearer to the one a bit related (I'd say between `cursor_at_end` and `fullscreen`), so it's easier to catch when looking for features.

File: frontend/ui/widget/inputdialog.lua [hunk 2]
```diff
@@ -347,16 +348,20 @@ function InputDialog:init()
                                     - vspan_after_input_text:getSize().h
                                     - buttons_container:getSize().h
                                     - keyboard_height
-        if self.fullscreen or text_height > available_height then
+                                    - Size.padding.large -- keep some padding around the dialog
```
Comment (hius07) [inline frontend/ui/widget/inputdialog.lua:351]: This affects Text editor (fullscreen mode) and reduces its window.

File: frontend/ui/widget/inputdialog.lua [hunk 3]
```diff
@@ -347,16 +348,20 @@ function InputDialog:init()
...
+        if self.fullscreen or self.use_available_height or text_height > available_height then
             -- Don't leave unusable space in the text widget, as the user could think
             -- it's an empty line: move that space in pads after and below (for centering)
+            local original_text_height = text_height
```
Comment (hius07) [inline frontend/ui/widget/inputdialog.lua:355]: Is new variable needed?

File: frontend/apps/reader/modules/readerbookmark.lua [hunk 4]
```diff
@@ -893,8 +893,9 @@ function ReaderBookmark:renameBookmark(item, from_highlight)
         description = "   " .. T(_("Page: %1"), bookmark.mandatory) .. "     " .. T(_("Time: %1"), bookmark.datetime),
         input = bookmark.text_orig,
         allow_newline = true,
-        cursor_at_end = false,
+        cursor_at_end = true,
```
Comment (hius07) [inline frontend/apps/reader/modules/readerbookmark.lua:896]: It's by default, so the line can be removed.

File: frontend/ui/widget/inputdialog.lua (hunk 3 above)
Comment (dgarrett) [reply frontend/ui/widget/inputdialog.lua]: Oh, I missed that `text_height` was local

File: frontend/ui/widget/inputdialog.lua [hunk 5]
```diff
@@ -347,7 +348,9 @@ function InputDialog:init()
                                     - vspan_after_input_text:getSize().h
                                     - buttons_container:getSize().h
                                     - keyboard_height
-        if self.fullscreen or text_height > available_height then
+                                    - (self.fullscreen and 0 or Size.padding.large) -- keep some padding around
```
Comment (hius07) [inline frontend/ui/widget/inputdialog.lua:351]: To keep current (good) behaviour everywhere, let's apply this new additional padding when `use_available_height` only.

Comment (poire-z) [reply frontend/ui/widget/inputdialog.lua:351]: Or just don't add any.
It feels you added it so "it looks good for me" - but as we add some additional padding in this branch, some people/device/dpi will get even more padding that what you added.
And with this additional padding, you'll possibly get one line less of what you'd get without - so, going against what you wished with this PR :)
And we were indeed fine with no padding previously, so we should still be fine without it.

Comment (dgarrett) [reply frontend/ui/widget/inputdialog.lua:351]: I'm fine with removing it. I just added the padding because "it looks good for me", as you said.

**Review by poire-z — APPROVED** [review]

**Pull request merged**